python/cudf/cudf/tests/test_list.py

```diff
@@ -107,6 +109,27 @@ def test_listdtype_hash():
     assert hash(a) != hash(c)
 
 
+@settings(deadline=None)
+@given(
+    st.builds(
+        lambda x: [x],
+        st.recursive(
+            st.integers(min_value=-100, max_value=100),
+            lambda c: st.lists(c, min_size=1, max_size=1)
```

Comment:
All entries in a list column must have the same dtype (which means that the generated children must all have the same "shape"), I couldn't figure out a nice way to express this. I can relax the singleton list aspect in the dict case.

Reply:
Aha. We might need to change our strategy. We probably want to do the following: first, choose a strategy `x` drawn from allowable leaf types (which could be structs, ints, floats, etc.). Then we would accept `lists(x)`, or `lists(lists(x))` up to arbitrary depth, but not `lists(x | lists(x))`.

I think what you want is something like:
```python
import hypothesis.strategies as st

leaf_sts = st.sampled_from([st.integers(), st.booleans()])

@st.composite
def nested_lists(draw, leaf_sts):
    leaf_st = draw(leaf_sts)
    nested_depth = draw(st.integers(min_value=0, max_value=3))
    nested_st = st.lists(leaf_st, max_size=3)
    for _ in range(nested_depth):
        nested_st = st.lists(nested_st, max_size=3)
    return draw(nested_st)

for _ in range(10):
    print(nested_lists(leaf_sts).example())
```

Reply:
I leave it as an exercise to define all supported leaf types and replace `leaf_st` in such a way that it supports struct leaf types -- feel free to ask if you want more help there.

Otherwise let's just define a few test cases manually and not use hypothesis, for the sake of ease of understanding and to reduce our (human, computer) time spent generating appropriate data strategies and running more tests than we probably need for adequate coverage here. 😄 I'd reserve this kind of task for fuzz testing and not unit tests.

Reply:
Thanks. I'll have a go next time round but I've gone for some hand-written tests this time.

(I think it would be worthwhile introducing more systematic fuzz-style testing here since when fixing a bug one has a test in mind, but that may not hit all the bugs, but this can happen separately).